### PR TITLE
NAS strategy (stage 2) - utils and bruteforce strategy

### PR DESCRIPTION
**Pull request by ultmaster**

### CODE REVIEW

File: nni/nas/strategy/bruteforce.py [hunk 1]
```diff
@@ -1,138 +1,232 @@
...
-class GridSearch(BaseStrategy):
+class GridSearch(Strategy):
     """
     Traverse the search space and try all the possible combinations one by one.
 
     Parameters
     ----------
-    shuffle : bool
-        Shuffle the order in a candidate list, so that they are tried in a random order. Default: true.
+    shuffle
```
Comment (liuzhe-lz) [inline nni/nas/strategy/bruteforce.py:30]: Does it present in v2.10? Seems it's incompatible with HPO grid search's design

File: nni/nas/strategy/bruteforce.py [hunk 2]
```diff
@@ -1,138 +1,232 @@
...
+        Shuffle the order in a candidate list, so that they are tried in a random order.
+        Currently, the implementation is a pseudo-random shuffle, which only shuffles the order of every 100 candidates.
+    dedup
```
Comment (liuzhe-lz) [inline nni/nas/strategy/bruteforce.py:33]: In what case one should turn off dedup?

Comment (ultmaster) [reply nni/nas/strategy/bruteforce.py:33]: For example when deduplication causes performance issues?

It won't happen in the wild I guess, but might happen in some benchmarks.

File: nni/nas/strategy/bruteforce.py (hunk 1 above)
Comment (ultmaster) [reply nni/nas/strategy/bruteforce.py:30]: Yes.

I think shuffled-grid-search is a nice feature and I prefer enabling it by default.

**Review by QuanluZhang — APPROVED** [review]

**Review by liuzhe-lz — APPROVED** [review]

**Pull request merged**